confidential-data-hub/hub/src/bin/confidential-data-hub/config.rs

```diff
@@ -0,0 +1,132 @@
+
+    pub credentials: Vec<Credential>,
+
+    pub socket: String,
+}
+
+impl Default for CdhConfig {
+    fn default() -> Self {
+        Self {
+            socket: DEFAULT_CDH_SOCKET_ADDR.into(),
+            kbc: KbsConfig::default(),
+            credentials: Vec::default(),
+        }
+    }
+}
+
+impl CdhConfig {
+    pub async fn init(config_path: &str) -> Result<Self> {
+        let mut config = Self::from_file(config_path).unwrap_or_else(|e| {
+            warn!("read config file {config_path} failed {e:?}, use a default config where `aa_kbc_params` = offline_fs_kbc::null.");
+            Self::default()
+        });
+
+        config.update_from_kernel_cmdline().await?;
+        Ok(config)
+    }
+
+    /// Load `CdhConfig` from a configuration file. Supported formats are all formats supported by the
+    /// `config` crate.
+    fn from_file(config_path: &str) -> Result<Self> {
+        let c = Config::builder()
+            .set_default("socket", DEFAULT_CDH_SOCKET_ADDR)?
+            .add_source(File::with_name(config_path))
+            .build()?;
+
+        let res = c.try_deserialize().context("invalid config").unwrap();
+        Ok(res)
+    }
+
+    /// all the resource ids can be from the kernel commandline in the following format:
+    /// ```shell
+    /// cdh.kbs_resources=<resource id 1>::<target path 1>,<resource id 2>::<target path 2>...
+    /// ```
+    ///
+    /// for example
+    /// ```shell
+    /// cdh.kbs_resources=kbs:///default/key/1::/run/temp1,kbs:///default/key/2::/run/temp2
+    /// ```
+    ///
+    /// It is supposed that all the `target path` should be with prefix
+    /// `/run/confidential-containers/cdh` or it will be treated as dangerous
+    /// path.
+    ///
+    /// TODO: delete this way after initdata mechanism could cover CDH's launch config.
+    async fn update_from_kernel_cmdline(&mut self) -> Result<()> {
+        let cmdline = fs::read_to_string("/proc/cmdline")
+            .await
+            .context("read kernel cmdline failed")?;
+        let kbs_resources = cmdline
+            .split_ascii_whitespace()
+            .find(|para| para.starts_with("cdh.kbs_resources="))
+            .unwrap_or("cdh.kbs_resources=")
+            .strip_prefix("cdh.kbs_resources=")
+            .expect("must have one")
+            .split(',')
+            .filter(|s| !s.is_empty())
+            .filter_map(|it| it.split_once("::"))
+            .map(|it| Credential {
+                resource_uri: it.0.to_owned(),
+                path: it.1.to_owned(),
+            });
+
+        self.credentials.extend(kbs_resources);
+        Ok(())
+    }
+}
+
+impl CdhConfig {
+    pub fn set_configuration_envs(&self) {
+        // KBS configurations
+        env::set_var(
+            "AA_KBC_PARAMS",
+            format!("{}::{}", self.kbc.name, self.kbc.url),
+        );
+        if let Some(kbs_cert) = &self.kbc.kbs_cert {
+            env::set_var("KBS_PUBLICKEY_CERT", kbs_cert);
+        }
+    }
```

Comment:
Since we are using AA as a library can we initialize the AttestationAgent instance with configuration data instead of setting and reading an env within the same binary?



Nit: KBS_PUBLICKEY_CERT can be shortened to KBS_CERT probably

Reply:
The attestation agent binary/lib is not used in CDH. I just use `aa_kbc_params` to get kbs and kbc.

Why I set env and read in a same binary, is
1. function call depth. `kbs` and `kbc` parameter are needed in a `kms`, a separate crate. If we want to deliver the as the function argument, a chain of functions would be needed to modify to deliver from config to the underneath, which would be messy. 
2. easy to use. From the point of a separate crate, an env would be good as kbc and kbs are simple strings that could
3. different crates. `kms` is a crate and `hub` another. With env, `kms` could both work itself and with hub.

Reply:
> Nit: KBS_PUBLICKEY_CERT can be shortened to KBS_CERT probably

Thanks for this. I fixed